File: lc/coreprep/implementation/practice_iterations/patterns_iterations_1/all_tasks_scheduling_order.cpp
#include <algorithm>
#include <iostream>
#include <queue>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;

using Graph = unordered_map<int, unordered_set<int>>;
using Degree = unordered_map<int, int>;

void all_tasks_scheduling(Graph &graph, Degree &indegree, deque<int> q,
                          vector<int> &curr, vector<vector<int>> &all) {
  if (!q.empty()) {
    deque<int> vs = q;

    while (!vs.empty()) {
      int v = vs.front();
      vs.pop_front();

      curr.push_back(v);
      deque<int> nextq = q;
      nextq.erase(remove(nextq.begin(), nextq.end(), v), nextq.end());

      for (int neighbour : graph[v]) {
        indegree[neighbour]--;

        if (indegree[neighbour] == 0) {
          nextq.push_back(neighbour);
        }
      }

      all_tasks_scheduling(graph, indegree, nextq, curr, all);

      curr.pop_back();
      for (int neighbour : graph[v]) {
        indegree[neighbour]++;
      }
    }
  }

  if (curr.size() == graph.size()) {
    all.push_back(curr);
  }
}

// O(V! x E) / O(V! x E)
vector<vector<int>> all_tasks_scheduling(Graph &graph, Degree &indegree) {
  deque<int> q;
  for (auto entry : indegree) {
    if (entry.second == 0) {
      q.push_back(entry.first);
    }
  }

  vector<vector<int>> res;
  vector<int> curr;
  all_tasks_scheduling(graph, indegree, q, curr, res);

  return res;
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, n, m, from, to;
  cin >> tt;

  while (tt--) {
    cin >> n >> m;

    Graph graph;
    Degree indegree;

    for (int i = 0; i < n; i++) {
      graph[i] = unordered_set<int>{};
      indegree[i] = 0;
    }

    for (int i = 0; i < m; i++) {
      cin >> from >> to;
      graph[from].insert(to);
      indegree[to]++;
    }

    auto res = all_tasks_scheduling(graph, indegree);
    for (auto r : res) {
      for (int e : r) {
        cout << e << " ";
      }
      cout << "\n";
    }
  }

  return 0;
}

File: lc/coreprep/implementation/practice_iterations/patterns_iterations_1/AllTasksSchedulingOrder.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

public class AllTasksSchedulingOrder {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();

        Map<Integer, Set<Integer>> graph = new HashMap<>();
        Map<Integer, Integer> indegree = new HashMap<>();

        for (int i = 0; i < n; i++) {
          graph.put(i, new HashSet<>());
          indegree.put(i, 0);
        }

        int m = in.nextInt();
        for (int i = 0; i < m; i++) {
          int from = in.nextInt();
          int to = in.nextInt();

          graph.get(from).add(to);
          indegree.put(to, indegree.get(to) + 1);
        }

        var res = allTasksSchedulingOrder(graph, indegree);
        for (var r : res) {
          for (int e : r) {
            System.out.printf("%d ", e);
          }
          System.out.println();
        }
      }
    }
  }

  // O(V! x E) / O(V! x E)
  public static List<List<Integer>>
  allTasksSchedulingOrder(Map<Integer, Set<Integer>> graph,
                          Map<Integer, Integer> indegree) {
    Queue<Integer> q = new ArrayDeque<>();
    for (var entry : indegree.entrySet()) {
      if (entry.getValue() == 0) {
        q.add(entry.getKey());
      }
    }

    List<List<Integer>> all = new ArrayList<>();
    List<Integer> curr = new ArrayList<>();

    allTasksSchedulingOrder(graph, indegree, q, curr, all);

    return all;
  }

  private static void allTasksSchedulingOrder(Map<Integer, Set<Integer>> graph,
                                              Map<Integer, Integer> indegree,
                                              Queue<Integer> q,
                                              List<Integer> curr,
                                              List<List<Integer>> all) {
    if (!q.isEmpty()) {
      for (int v : q) {
        curr.add(v);

        Queue<Integer> nextq = new ArrayDeque<>(q);
        nextq.remove(v);

        for (int neighbour : graph.get(v)) {
          indegree.put(neighbour, indegree.get(neighbour) - 1);

          if (indegree.get(neighbour) == 0) {
            nextq.add(neighbour);
          }
        }

        allTasksSchedulingOrder(graph, indegree, nextq, curr, all);

        curr.remove(curr.size() - 1);
        for (int neighbour : graph.get(v)) {
          indegree.put(neighbour, indegree.get(neighbour) + 1);
        }
      }
    }

    if (curr.size() == graph.size()) {
      all.add(new ArrayList<>(curr));
    }
  }
}